### PR TITLE
MCOL-4394 __float128 related code had been moved into a separate file

### PR DESCRIPTION
**Pull request by drrtuy**
Trim to double and to long double conversions for Decimal

### CODE REVIEW

**Review by abarkov — APPROVED** [review]

File: datatypes/mcs_decimal.h [hunk 1]
```diff
@@ -598,6 +569,35 @@ class VDecimal: public TSInt128
...
+    inline operator double() const
+    {
+        return toDouble();
+    }
+
+    inline long double toLongDouble() const
+    {
+        datatypes::TFloat128 y(s128Value);
+        return static_cast<long double>(y);
+    }
```
Comment (tntnatbry) [inline datatypes/mcs_decimal.h:599]: The semantics of `toLongDouble()` should be the same as `toDouble()`, i.e., we should use the `scale` here to construct `TFloat128`.

Also, we should also add `operator long double()` as a member function here.

File: datatypes/mcs_int128.h [hunk 2]
```diff
@@ -203,6 +180,36 @@ class TSInt128
       return s128Value == static_cast<int128_t>(x);
     }
 
+    inline operator double() const
+    {
+      return toDouble();
+    }
+
+    inline long double toDouble() const
```
Comment (tntnatbry) [inline datatypes/mcs_int128.h:188]: Return type here should be `double` and not `long double`.

File: datatypes/mcs_int128.h [hunk 3]
```diff
@@ -203,6 +180,36 @@ class TSInt128
...
+    {
+      return static_cast<double>(s128Value);
```
Comment (tntnatbry) [inline datatypes/mcs_int128.h:190]: I think this conversion can overflow, this should be `return static_cast<double>(TFloat128(s128Value))`.

File: datatypes/mcs_int128.h [hunk 4]
```diff
@@ -203,6 +180,36 @@ class TSInt128
...
+    }
+
+    inline operator long double() const
+    {
+      return toLongDouble();
+    }
+
+    inline long double toLongDouble() const
+    {
+      return static_cast<long double>(s128Value);
```
Comment (tntnatbry) [inline datatypes/mcs_int128.h:200]: I think this conversion can overflow, this should be `return static_cast<long double>(TFloat128(s128Value))`.

File: utils/rowgroup/rowaggregation.cpp [hunk 5]
```diff
@@ -2045,8 +2045,9 @@ void RowAggregation::doStatistics(const Row& rowIn, int64_t colIn, int64_t colOu
         case execplan::CalpontSystemCatalog::UDECIMAL:  // handle scale later
             if (LIKELY(fRowGroupIn.getColumnWidth(colIn) == datatypes::MAXDECIMALWIDTH))
             {
-                int128_t* val128InPtr = rowIn.getBinaryField<int128_t>(colIn);
-                valIn = Dec::getLongDoubleFromWideDecimal(*val128InPtr);
+                // To save from unaligned memory
+                datatypes::TSInt128 val128In(rowIn.getBinaryField<int128_t>(colIn));
+                valIn =  static_cast<long double>(val128In.toTFloat128());
```
Comment (tntnatbry) [inline utils/rowgroup/rowaggregation.cpp:2050]: The `.toTFloat128()` call here is unnecessary.

**Comment by tntnatbry** [comment]
@drrtuy Please look at the review comments I added to this PR.

File: datatypes/mcs_decimal.h (hunk 1 above)
Comment (drrtuy) [reply datatypes/mcs_decimal.h:599]: To be honest we don't need this function at this point so I'll remove it.

File: datatypes/mcs_int128.h (hunk 2 above)
Comment (drrtuy) [reply datatypes/mcs_int128.h:188]: Good catch. I wonder how did this pass through the compiler.

File: datatypes/mcs_int128.h (hunk 4 above)
Comment (drrtuy) [reply datatypes/mcs_int128.h:200]: This is a very good question. As you can see I didn't change the behavior so I trusted whatever logic existed before I wrote this. And I believe the logic had been introduced by you in the first place.

File: utils/rowgroup/rowaggregation.cpp (hunk 5 above)
Comment (drrtuy) [reply utils/rowgroup/rowaggregation.cpp:2050]: Not exactly. TSInt128::operator long double() shouldn't be used here b/c I wish  to remove all casts from int128 -> real types b/c they doesn't look right. TFloat128 should be used instead. We'll have extra 16 byte asignment step here but we won't be casting from int128 into a real type.

**Pull request merged**